fabric-data-attachment-api-v1/src/main/java/net/fabricmc/fabric/api/attachment/v1/AttachmentTarget.java

```diff
@@ -0,0 +1,121 @@
+ *
+ * <p>Fabric implements this on {@link Entity}, {@link BlockEntity}, {@link ServerWorld} and {@link WorldChunk} via mixin.</p>
+ */
+public interface AttachmentTarget {
+	/**
+	 * Gets the data associated with the given {@link Attachment}. If it doesn't exist yet,
+	 * it is generated using {@link Attachment#initializer()}.
+	 *
+	 * @param attachment the attachment
+	 * @param <A> the type of the data
+	 * @return the attached data
+	 */
+	default <A> A getAttached(Attachment<A> attachment) {
+		throw new UnsupportedOperationException("Implemented via mixin");
+	}
+
+	/**
+	 * Sets the data associated with the given {@link Attachment}. Passing {@code null} removes the data.
+	 *
+	 * @param attachment the attachment
+	 * @param value the new value
+	 * @param <A> the type of the data
+	 * @return the previous data
+	 */
+	@Nullable
+	default <A> A setAttached(Attachment<A> attachment, @Nullable A value) {
+		throw new UnsupportedOperationException("Implemented via mixin");
+	}
+
+	/**
+	 * Tests whether the given {@link Attachment} has any associated data.
+	 *
+	 * @param attachment the attachment
+	 * @return whether there is associated data
+	 */
+	default boolean hasAttached(Attachment<?> attachment) {
+		throw new UnsupportedOperationException("Implemented via mixin");
+	}
+
+	/**
+	 * Removes any data associated with the given {@link Attachment}. Equivalent to calling {@link #setAttached(Attachment, Object)}
+	 * with {@code null}.
+	 *
+	 * @param attachment the attachment
+	 * @param <A> the type of the data
+	 * @return the previous data
+	 */
+	@Nullable
+	default <A> A removeAttached(Attachment<A> attachment) {
+		return setAttached(attachment, null);
+	}
+
+	/**
+	 * Modifies the data associated with the given {@link Attachment}. Functionally the same as calling {@link #getAttached(Attachment)},
+	 * applying the modifier, then calling {@link #setAttached(Attachment, Object)} with the result.
+	 *
+	 * @param attachment the attachment
+	 * @param modifier the operation to apply to the current data
+	 * @param <A> the type of the data
+	 * @return the previous data
+	 */
+	@Nullable
+	default <A> A modifyAttached(Attachment<A> attachment, UnaryOperator<A> modifier) {
+		return setAttached(attachment, modifier.apply(getAttached(attachment)));
+	}
+
+	/**
+	 * Writes all the attached data to NBT.
+	 *
+	 * @param nbt the NBT to write to
+	 */
+	default void writeAttachmentsToNbt(NbtCompound nbt) {
+		throw new UnsupportedOperationException("Implemented via mixin");
+	}
+
+	/**
+	 * Reads all the attached data from NBT.
+	 *
+	 * @param nbt the NBT to read from
+	 */
+	default void readAttachmentsFromNbt(NbtCompound nbt) {
+		throw new UnsupportedOperationException("Implemented via mixin");
+	}
+
+	/**
+	 * @return whether the {@link AttachmentTarget} has any attachment data to serialize
+	 */
+	default boolean hasSerializableAttachments() {
+		throw new UnsupportedOperationException("Implemented via mixin");
+	}
```

Comment:
Do these need to be part of the public API?

Reply:
I wasn't entirely sure how to make this impl-only, but I've figured out a way.